lib/hawkular/prometheus/prometheus_api.rb

```diff
@@ -0,0 +1,50 @@
+require 'hawkular/base_client'
+require 'hawkular/inventory/entities'
+require 'ostruct'
+
+module Hawkular::Prometheus
+  class AlerterClient < Hawkular::BaseClient
+    def initialize(entrypoint, credentials = {}, options = {})
+      entrypoint = normalize_entrypoint_url entrypoint, 'hawkular/alerter'
+      @entrypoint = entrypoint
+
+      super(entrypoint, credentials, options)
+    end
+
+    def prometheus_entrypoint
+      rest_client('/prometheus/endpoint').get
+    end
+  end
+
+  # Interface to talk with the Prometheus server used for Middleware Manager
+  # @param entrypoint [String] base url of Hawkular Services
+  class Client < Hawkular::BaseClient
+    def initialize(entrypoint, credentials = {}, options = {})
+      prometheus_entrypoint = AlerterClient.new(entrypoint, credentials, options).prometheus_entrypoint
+      entrypoint = normalize_entrypoint_url prometheus_entrypoint, 'api/v1'
+      @entrypoint = entrypoint
+
+      super(entrypoint, credentials, options)
+    end
+
+    def query(metrics: [], time: nil)
+      query = metrics.map(&:expression).join(' or ')
+      response = http_get "/query?time=#{time}&query=#{query}"
+      results = response['data']['result']
+      attach_metric_in_result(results, metrics)
+    end
+
+    def query_range(metrics: [], starts: nil, ends: nil, step: nil)
+      query = metrics.map(&:expression).join(' or ')
+      response = http_get "/query_range?start=#{starts}&end=#{ends}&step=#{step}&query=#{query}"
+      results = response['data']['result']
+      attach_metric_in_result(results, metrics)
+    end
+
+    private
+
+    def attach_metric_in_result(results, metrics)
+      results.each_with_index { |result, index| result['metric'] = metrics.fetch(index) }
```

Comment:
This seems... weird. First weird thing is that we are returning hashes instead of objects, which is different from `Inventory`, for example. Second, it seems extremely contrived to do this, but to be honest I can not think of a better option right now, maybe something with `Enumerable#zip`?

Reply:
The feature of this is to attach the metric model into the results which is what is needed for the provider's logic.